File: 백준/Gold/21315. 카드 섞기/카드 섞기.py
import sys

def magic(k, card):
    arr = []
    for i in range(1, k + 2):
        arr.extend(card[:len(card) - 2 ** (k - i + 1)][::-1])
        card = card[len(card) - 2 ** (k - i + 1):]
    arr.extend(card[::-1])
    return arr[::-1]

N = int(input())
after = [*map(int, input().split())]
card = [k for k in range(1, N + 1)]
for i in range(1, 10):
    if N < 2 ** i: break
    mres = magic(i, card)
    for j in range(1, 10):
        if N < 2 ** j: break
        if magic(j, mres) == after:
            print(i, j)
            sys.exit()

File: 백준/Gold/21315. 카드 섞기/카드 섞기.java
import java.io.*;
import java.util.*;

public class Main {

    public static int N;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        N = Integer.parseInt(br.readLine());
        String after = br.readLine();
        int[] card = new int[N];
        for (int i = 1; i <= N; i++) {
            card[i - 1] = i;
        }

        for (int i = 1; i < 10; i++) {
            if (N < Math.pow(2, i)) {
                break;
            }
            String mres = magic(i, card);
            for (int j = 1; j < 10; j++) {
                if (N < Math.pow(2, j)) {
                    break;
                }
                if (after.equals(magic(j, convert(mres)))) {
                    System.out.println(i + " " + j);
                    return;
                }
            }
        }
    }

    public static String magic(int k, int[] card) {
        String res = "";
        for (int i = 1; i < k + 2; i++) {
            for (int j = card.length - (int) Math.pow(2, k - i + 1) - 1; j >= 0; j--) {
                if (res.length() == 0) {
                    res += card[j];
                } else {
                    res = card[j] + " " + res;
                }
            }
            int[] tmp = new int[(int) Math.pow(2, k - i + 1)];
            int idx = 0;
            for (int j = card.length - (int) Math.pow(2, k - i + 1); j < card.length; j++) {
                tmp[idx++] = card[j];
            }
            card = tmp;
        }
        return card[0] + " " + res;
    }

    public static int[] convert(String nums) {
        return Arrays.stream(nums.split(" ")).mapToInt(Integer::parseInt).toArray();
    }

}